### PR TITLE
Use MeanOp instead of GlobalAvgPool2dOp

### PR DESCRIPTION
**Pull request by mvasiljevicTT**
### Ticket
[CNN Fusing Issue - use mean instead of global_avg_pool](https://github.com/tenstorrent/tt-mlir/issues/5511)
[CNN Fusing Issue - enable-global-avg-pool-pattern on by default](https://github.com/tenstorrent/tt-mlir/issues/5427)
[CNN EIO Issue - commute permute with reduction ops](https://github.com/tenstorrent/tt-mlir/issues/4698)

### Problem description
Due to this [tt-metal issue](https://github.com/tenstorrent/tt-metal/issues/31065) we should change current patterns to use mean op instead of global_avg_pool2d. 

Additinaly, in efficientnet model in torch-xla global_avg_pool is between convolutions that are converted to conv2d ops with NCHW <-> NHWC permutation, making global_avg_pool, now mean, surrounded by inverse permutations. 

This should be improved like this:
```
perm
|
mean<dims>      ->     mean<perm(dims)>
|
perm^-1
```

### What's changed
- Changed  `GlobalAveragePoolingPattern` to use `MeanOp` instead of `GlobalAvgPool2dOp`.
- Removed `enable-global-avg-pool-pattern` flag as that pattern is now always enabled.
- Added `ReduceCommutePatterns` in `EraseInverseOps` pass for commuting reduce ops that have `keepdim=true` with permute as:

```
perm
|
reduce<dims>   ->    reduce<perm(dims)>
                     |
                     perm
```
and
```
                     perm
                     |
reduce<dims>   ->    reduce<perm^-1(dims)>
|
perm
```

### Checklist
- [x] New/Existing tests provide coverage for changes

### CODE REVIEW

File: lib/Dialect/TTIR/Transforms/EraseInverseOps/ReduceCommutePatterns.cpp [hunk 1]
```diff
@@ -0,0 +1,191 @@
+// SPDX-FileCopyrightText: (c) 2025 Tenstorrent AI ULC
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "ttmlir/Dialect/TTIR/Transforms/EraseInverseOps/EraseInverseOps.h"
+
+#include "mlir/IR/BuiltinTypes.h"
```
Comment (vmilosevic) [inline lib/Dialect/TTIR/Transforms/EraseInverseOps/ReduceCommutePatterns.cpp:8]: :warning: **llvm\-include\-order** :warning:
\#includes are not sorted properly
```suggestion
#include "mlir/IR/BuiltinAttributes.h"
#include "mlir/IR/BuiltinTypes.h"
```

File: lib/Dialect/TTIR/Transforms/EraseInverseOps/ReduceCommutePatterns.cpp [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+#include "ttmlir/Dialect/TTCore/IR/Utils.h"
+#include "ttmlir/Dialect/TTIR/IR/TTIROps.h"
+#include "ttmlir/Dialect/TTIR/Utils/Utils.h"
+#include <flatbuffers/array.h>
```
Comment (vmilosevic) [inline lib/Dialect/TTIR/Transforms/EraseInverseOps/ReduceCommutePatterns.cpp]: :warning: **llvm\-include\-order** :warning:
\#includes are not sorted properly
```suggestion

```

File: lib/Dialect/TTIR/Transforms/EraseInverseOps/ReduceCommutePatterns.cpp [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/Support/raw_ostream.h"
+#include "mlir/IR/BuiltinAttributes.h"
```
Comment (vmilosevic) [inline lib/Dialect/TTIR/Transforms/EraseInverseOps/ReduceCommutePatterns.cpp]: :warning: **llvm\-include\-order** :warning:
\#includes are not sorted properly
```suggestion
#include <flatbuffers/array.h>
```

File: lib/Dialect/TTIR/Transforms/EraseInverseOps/ReduceCommutePatterns.cpp [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+                              op.getType().getEncoding());
+
+    return utils::createDPSOp<ReduceOpType>(
+        rewriter, op->getLoc(), newReduceType, newInput, op.getKeepDimAttr(),
+        newDimArgAttrs);
+  }
+
+  bool isCommuteUpwardsViable(ReduceOpType op, PermuteOp) const override {
+    // Commuting a permute through a reduce is only viable if keepdim = true
+    return op.getKeepDim();
```
Comment (odjuricicTT) [inline lib/Dialect/TTIR/Transforms/EraseInverseOps/ReduceCommutePatterns.cpp]: Why is this the case? On first thought it seems like it could be done with keep dim false as well? We might not need it tho.

File: lib/Dialect/TTIR/Transforms/EraseInverseOps/ReduceCommutePatterns.cpp [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+  // The reduce dimensions are transformed by applying the inverse permutation
+  // Input and output shapes of reduce are permuted since permute op now come
+  // before reduce
+
+  void performCommuteUpwardsRewrite(ReduceOpType op, PermuteOp permuteUser,
+                                    PatternRewriter &rewriter) const override {
+    ArrayRef<int64_t> permutation = permuteUser.getPermutation();
+
+    PermuteOp newPerm =
+        createNewPermuteOp(op.getInput(), permutation, rewriter, op->getLoc());
```
Comment (odjuricicTT) [inline lib/Dialect/TTIR/Transforms/EraseInverseOps/ReduceCommutePatterns.cpp]: Better to use the loc of the permuteUser?

File: lib/Dialect/TTIR/Transforms/EraseInverseOps/ReduceCommutePatterns.cpp [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+
+    ReduceOpType newReduce = createNewReduceOpWithPermutedDims(
+        op, newPerm->getResult(0), permutation, rewriter,
+        /*inverseDimPermute=*/true);
+
+    rewriter.replaceAllOpUsesWith(permuteUser, newReduce);
```
Comment (odjuricicTT) [inline lib/Dialect/TTIR/Transforms/EraseInverseOps/ReduceCommutePatterns.cpp]: What happens if the initial reduce op has multiple users?

File: test/ttmlir/Dialect/TTIR/Transforms/EraseInverseOps/commute_permute_reduce.mlir [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    // CHECK-NOT: "ttir.permute"
+    %0 = tensor.empty() : tensor<12x1152x7x7xbf16>
+    %1 = "ttir.permute"(%arg0, %0) <{permutation = array<i64: 0, 3, 1, 2>}> : (tensor<12x7x7x1152xbf16>, tensor<12x1152x7x7xbf16>) -> tensor<12x1152x7x7xbf16>
+    %2 = tensor.empty() : tensor<12x1152x1x7xi32>
+    %3 = "ttir.argmax"(%1, %2) <{dim_arg = [2 : i32], keep_dim = true}> : (tensor<12x1152x7x7xbf16>, tensor<12x1152x1x7xi32>) -> tensor<12x1152x1x7xi32>
+    %4 = tensor.empty() : tensor<12x1x7x1152xi32>
+    %5 = "ttir.permute"(%3, %4) <{permutation = array<i64: 0, 2, 3, 1>}> : (tensor<12x1152x1x7xi32>, tensor<12x1x7x1152xi32>) -> tensor<12x1x7x1152xi32>
+    return %5 : tensor<12x1x7x1152xi32>
+  }
+}
```
Comment (odjuricicTT) [inline test/ttmlir/Dialect/TTIR/Transforms/EraseInverseOps/commute_permute_reduce.mlir]: Can you add a test where the reduciton op has multiple permute users?

And one where permute cannot commute because of keep_dim.
And on where permute cannot commute because of different users.

File: lib/Dialect/TTIR/Transforms/TTIRFusing.cpp [hunk 8]
```diff
@@ -1564,15 +1564,10 @@ class AveragePoolingWithPoolingDenominatorFusionPattern
...
 //
 // Supports 4D input tensors with reduction over spatial dimensions (height and
 // width).
-//
-// Input tensor layout is NCHW, while AvgPool2dOp expects NHWC layout, so
-// PermuteOps are added before and after the AvgPool2dOp to convert between the
-// two layouts, and a ReshapeOp is added after the permutes if the
-// keepDim attribute of the original SumOp is false.
 
 class GlobalAveragePoolingPattern : public mlir::OpRewritePattern<MultiplyOp> {
```
Comment (odjuricicTT) [inline lib/Dialect/TTIR/Transforms/TTIRFusing.cpp]: Should the pattern be renamed as well?

File: test/ttmlir/Dialect/TTIR/fusing/scaled_sum_to_mean_fusing.mlir [hunk 9]
```diff
@@ -1,4 +1,4 @@
-// RUN: ttmlir-opt --canonicalize --ttir-fusing="ttnn-enable-global-avg-pool-pattern=true" -o %t %s
+// RUN: ttmlir-opt --canonicalize --ttir-fusing -o %t %s
```
Comment (odjuricicTT) [inline test/ttmlir/Dialect/TTIR/fusing/scaled_sum_to_mean_fusing.mlir:1]: Also rename the test if the pattern gets renamed.

**Review by odjuricicTT — COMMENTED** [review]
Great, tho a few comment regarding some corner cases.

File: lib/Dialect/TTIR/Transforms/EraseInverseOps/ReduceCommutePatterns.cpp (hunk 4 above)
Comment (mvasiljevicTT) [reply lib/Dialect/TTIR/Transforms/EraseInverseOps/ReduceCommutePatterns.cpp]: It could be done , but it is not that straightforward and we don't need it so I skipped it for now. I change the comment.

File: lib/Dialect/TTIR/Transforms/EraseInverseOps/ReduceCommutePatterns.cpp (hunk 5 above)
Comment (mvasiljevicTT) [reply lib/Dialect/TTIR/Transforms/EraseInverseOps/ReduceCommutePatterns.cpp]: Yup, that was a typo, changed now :D

File: lib/Dialect/TTIR/Transforms/EraseInverseOps/ReduceCommutePatterns.cpp (hunk 6 above)
Comment (mvasiljevicTT) [reply lib/Dialect/TTIR/Transforms/EraseInverseOps/ReduceCommutePatterns.cpp]: If they are not all same permute it is not run at all as it is *not favorable*. Otherwise fixed now.

File: test/ttmlir/Dialect/TTIR/Transforms/EraseInverseOps/commute_permute_reduce.mlir (hunk 7 above)
Comment (mvasiljevicTT) [reply test/ttmlir/Dialect/TTIR/Transforms/EraseInverseOps/commute_permute_reduce.mlir]: Added missing cases and reorganized to `.../CommuteDownwards/commute_permute_reduce.mlir` and `.../CommuteUpwards/commute_permute_reduce.mlir`

File: lib/Dialect/TTIR/Transforms/TTIRFusing.cpp (hunk 8 above)
Comment (mvasiljevicTT) [reply lib/Dialect/TTIR/Transforms/TTIRFusing.cpp]: Renamed to `ScaledSumToMeanPattern`

File: test/ttmlir/Dialect/TTIR/fusing/scaled_sum_to_mean_fusing.mlir (hunk 9 above)
Comment (mvasiljevicTT) [reply test/ttmlir/Dialect/TTIR/fusing/scaled_sum_to_mean_fusing.mlir:1]: renamed tests to `scaled_sum_to_mean.*`

File: lib/Dialect/TTIR/Transforms/EraseInverseOps/ReduceCommutePatterns.cpp [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+    PermuteOp newPerm = createNewPermuteOp(op.getInput(), permutation, rewriter,
+                                           permuteUser->getLoc());
+
+    ReduceOpType newReduce = createNewReduceOpWithPermutedDims(
+        op, newPerm->getResult(0), permutation, rewriter,
+        /*inverseDimPermute=*/true);
+    SmallVector<Operation *> users(op->getUsers());
+    for (auto *user : users) {
+      assert(checkIdenticalTms(permuteUser, user) &&
+             "shouldCommute should have ensured this is true");
```
Comment (odjuricicTT) [inline lib/Dialect/TTIR/Transforms/EraseInverseOps/ReduceCommutePatterns.cpp:52]: What is shouldCommute in this context?

File: lib/Dialect/TTIR/Transforms/EraseInverseOps/ReduceCommutePatterns.cpp [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+public:
+  using TTIRCommuteOpRewritePattern<
+      PermuteOp, ReduceOpType, commuteDirection>::TTIRCommuteOpRewritePattern;
+
+  // Consider the following IR pseudocode:
+  // %0 = reduce(%arg0, %output) {keep_dim = true, dim_arg = [2: i32, 3: i32]}
+  // %1 = permute(%0, %output1) <{permutation = array<i64: 0, 2, 3, 1>}>
+  //
+  // This method will transform this into:
+  // %0 = permute(%arg0, %output0) <{permutation = array<i64: 0, 3, 1, 2>}>
```
Comment (mtopalovicTT) [inline lib/Dialect/TTIR/Transforms/EraseInverseOps/ReduceCommutePatterns.cpp]: Is this always safe to do? I cannot wrap my head around it.

File: lib/Dialect/TTIR/Transforms/EraseInverseOps/ReduceCommutePatterns.cpp (hunk 10 above)
Comment (mvasiljevicTT) [reply lib/Dialect/TTIR/Transforms/EraseInverseOps/ReduceCommutePatterns.cpp:52]: It is referring to any of `isCommuteViable/Favorable`. 
I change the messages in all EIO rewrites to `"isCommuteUpwardsViable/Favorable should have ensured all users are identical TMs"` (it is upwards in all cases)

File: lib/Dialect/TTIR/Transforms/EraseInverseOps/ReduceCommutePatterns.cpp (hunk 11 above)
Comment (mvasiljevicTT) [reply lib/Dialect/TTIR/Transforms/EraseInverseOps/ReduceCommutePatterns.cpp]: I changed the comment now, both permutations should be `<0, 2, 3, 1>`.
It is always equivalent - for tensor with shape `[a,b,c,d]` if we first reduce it by `b` (`dim=1`) and then permute the result with any permutation e.g. `[d,a,c,1]` it is same as if we first permute it to `[d,a,c,b]` and then reduce by `b` (`dim=perm^-1(1)`). 
It is same for reducing by multiple dims.

File: lib/Dialect/TTIR/Transforms/EraseInverseOps/ReduceCommutePatterns.cpp [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+public:
+  using TTIRCommuteOpRewritePattern<
+      PermuteOp, ReduceOpType, commuteDirection>::TTIRCommuteOpRewritePattern;
+
+  // Consider the following IR pseudocode:
+  // %0 = reduce(%arg0, %output) {keep_dim = true, dim_arg = [2: i32, 3: i32]}
+  // %1 = permute(%0, %output1) <{permutation = array<i64: 0, 2, 3, 1>}>
+  //
+  // This method will transform this into:
+  // %0 = permute(%arg0, %output0) <{permutation = array<i64: 0, 2, 3, 1>}>
```
Comment (mtopalovicTT) [inline lib/Dialect/TTIR/Transforms/EraseInverseOps/ReduceCommutePatterns.cpp]: %output0 -> %output

File: lib/Dialect/TTIR/Transforms/EraseInverseOps/ReduceCommutePatterns.cpp [hunk 13]
```diff
@@ -0,0 +1,196 @@
...
+  }
+
+private:
+  PermuteOp createNewPermuteOp(Value input, ArrayRef<int64_t> permutation,
+                               PatternRewriter &rewriter, Location loc) const {
+    // Apply permutation on input type to create output type
+    auto inputType = cast<RankedTensorType>(input.getType());
+    SmallVector<int64_t> inputShape(inputType.getShape());
+    SmallVector<int64_t> outputShape = ttmlir::utils::applyPermutation(
+        ArrayRef<int64_t>(inputShape), permutation);
```
Comment (mtopalovicTT) [inline lib/Dialect/TTIR/Transforms/EraseInverseOps/ReduceCommutePatterns.cpp]: I don't think you need explicit cast to array ref here. Same on other places.

File: lib/Dialect/TTIR/Transforms/EraseInverseOps/ReduceCommutePatterns.cpp [hunk 14]
```diff
@@ -0,0 +1,196 @@
...
+
+  void performCommuteUpwardsRewrite(ReduceOpType op, PermuteOp permuteUser,
+                                    PatternRewriter &rewriter) const override {
+    ArrayRef<int64_t> permutation = permuteUser.getPermutation();
+
+    PermuteOp newPerm = createNewPermuteOp(op.getInput(), permutation, rewriter,
+                                           permuteUser->getLoc());
+
+    ReduceOpType newReduce = createNewReduceOpWithPermutedDims(
+        op, newPerm->getResult(0), permutation, rewriter,
```
Comment (mtopalovicTT) [inline lib/Dialect/TTIR/Transforms/EraseInverseOps/ReduceCommutePatterns.cpp]: newPerm.getResult()

File: lib/Dialect/TTIR/Transforms/EraseInverseOps/ReduceCommutePatterns.cpp [hunk 15]
```diff
@@ -0,0 +1,196 @@
...
+      if (op.isDpsInit(&op->getOpOperand(i))) {
+        continue;
+      }
+      if (checkIdenticalTms(op->getOperand(i).getDefiningOp(),
+                            permuteOperand) ||
+          ttcore::valueTracesToConstantArgs(op->getOperand(i))) {
+        continue;
+      }
+      return false;
+    }
```
Comment (mtopalovicTT) [inline lib/Dialect/TTIR/Transforms/EraseInverseOps/ReduceCommutePatterns.cpp]:
```suggestion
    auto dps = mlir::cast<mlir::DestinationStyleOpInterface>(op.getOperation());
    for (mlir::OpOperand *operand : dps.getDpsInputOperands()) {
      mlir::Value operandValue = operand->get();
      if (checkIdenticalTms(operandValue.getDefiningOp(),
                            permuteOperand) ||
          ttcore::valueTracesToConstantArgs(operandValue)) {
        continue;
      }
      return false;
    }
```

**Review by mtopalovicTT — APPROVED** [review]

File: lib/Dialect/TTIR/Transforms/EraseInverseOps/ReduceCommutePatterns.cpp (hunk 12 above)
Comment (mvasiljevicTT) [reply lib/Dialect/TTIR/Transforms/EraseInverseOps/ReduceCommutePatterns.cpp]: removed output arg from examples in comments now

File: lib/Dialect/TTIR/Transforms/EraseInverseOps/ReduceCommutePatterns.cpp (hunk 13 above)
Comment (mvasiljevicTT) [reply lib/Dialect/TTIR/Transforms/EraseInverseOps/ReduceCommutePatterns.cpp]: thanks, refactored now

File: lib/Dialect/TTIR/Transforms/EraseInverseOps/ReduceCommutePatterns.cpp (hunk 14 above)
Comment (mvasiljevicTT) [reply lib/Dialect/TTIR/Transforms/EraseInverseOps/ReduceCommutePatterns.cpp]: changed now

File: lib/Dialect/TTIR/Transforms/EraseInverseOps/ReduceCommutePatterns.cpp (hunk 15 above)
Comment (mvasiljevicTT) [reply lib/Dialect/TTIR/Transforms/EraseInverseOps/ReduceCommutePatterns.cpp]: that's cool, changed now
